src/ruby/pb/test/xds_client.rb

```diff
@@ -0,0 +1,193 @@
+    index = 0
+    $watchers_mutex.synchronize do
+      index = $watchers.length
+      # Add a watcher for this range of id's
+      $watchers << {
+        "start_id" => start_id,
+        "end_id" => end_id,
+        "rpcs_by_peer" => Hash.new(0),
+        "rpcs_needed" => req['num_rpcs'],
+        "no_remote_peer" => 0
+      }
+    end
+    # wait until all RPCs are done, or we are past timeout
+    while ($watchers[index]['rpcs_needed'] > 0) &&
+          (Time.now.to_f < finish_time)
+      sleep(0.1)
+    end
+    LoadBalancerStatsResponse.new(
+      rpcs_by_peer: $watchers[index]['rpcs_by_peer'],
+      num_failures: $watchers[index]['no_remote_peer'] +
+      $watchers[index]['rpcs_needed']
+    );
+  end
+end
+
+# send 1 rpc every 1/qps second
+def run_test_loop(stub, duration_per_query)
+  include Grpc::Testing
+  req = SimpleRequest.new()
+
+  while true
+    start = Time.now.to_f
+    begin
+      resp = stub.unary_call(req)
+      remote_peer = resp.hostname
+    rescue GRPC::Unavailable => e
+      remote_peer = ""
+    end
+    finish = Time.now.to_f
+    delay = duration_per_query - (finish - start)
+    if delay > 0
+      sleep(delay)
+    else
+      GRPC.logger.info("ruby xds: warning, rpc takes too long to finish. " \
+                       "If you consistently see this, the qps is too high.")
+    end
+    $current_request_id += 1
+    $watchers_mutex.synchronize do
+      $watchers.each do |watcher|
+        if (watcher['start_id'] <= $current_request_id) &&
+           ($current_request_id < watcher['end_id'])
+          watcher['rpcs_needed'] -= 1
+          if remote_peer.strip.empty?
+            watcher['no_remote_peer'] += 1
+          else
+            watcher['rpcs_by_peer'][remote_peer] += 1
+          end
+        end
+      end
+    end
+  end
+end
+
+# Args is used to hold the command line info.
+Args = Struct.new(:server, :stats_port, :qps)
+
+# validates the command line options, returning them as a Hash.
+def parse_args
+  args = Args.new
+  OptionParser.new do |opts|
+    opts.on('--server SERVER_HOST', 'server hostname') do |v|
+      GRPC.logger.info("ruby xds: server address is #{v}")
+      args['server'] = v
+    end
+    opts.on('--stats_port STATS_PORT', 'stats port') do |v|
+      GRPC.logger.info("ruby xds: stats port is #{v}")
+      args['stats_port'] = v
+    end
+    opts.on('--qps QPS', 'qps') do |v|
+      GRPC.logger.info("ruby xds: qps is #{v}")
+      args['qps'] = v
+    end
+  end.parse!
+  args
+end
+
+def main
+  opts = parse_args
+
+  # This server hosts the LoadBalancerStatsService
+  host = "0.0.0.0:#{opts['stats_port']}"
+  s = GRPC::RpcServer.new
+  s.add_http2_port(host, :this_port_is_insecure)
+  s.handle(TestTarget)
+  server_thread = Thread.new {
+    s.run_till_terminated
+  }
+
+  # The client just sends unary rpcs continously in a regular interval
+  stub = create_stub(opts)
+  duration_per_query = (1.0 / opts['qps'].to_f)
+  client_thread = Thread.new {
```

Comment:
per the spec I think we need to add an `--num_channels` option here, and then create `--num_channels` stubs, eaching having their own one of these threads

Reply:
Added